### PR TITLE
coconut: init at 1.4.3

### PR DESCRIPTION
**Pull request by fabianhjr**
Add [Coconut](http://coconut-lang.org/) interpreter.

Includes a patch that affected building: https://github.com/evhub/coconut/pull/560

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/python-modules/cPyparsing/default.nix [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{lib, buildPythonPackage, fetchPypi, cython}:
+
+buildPythonPackage rec {
+  pname = "cPyparsing";
+  version = "2.4.5.0.1.1";
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cPyparsing/default.nix:5]: What the hell is this version number. I've never seen so many digits.

File: pkgs/development/python-modules/cPyparsing/default.nix [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1kcr8v9yx9whbdn7c7rnfg4b6bk3agpa1wblhr2y6n93r8a4bxwy";
+  };
+
+  buildInputs = [ cython ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cPyparsing/default.nix:12]: Is this required at runtime? If so use propagatedBuildInputs.

File: pkgs/development/python-modules/cPyparsing/default.nix [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+  meta = with lib; {
+    homepage = "https://github.com/evhub/cpyparsing";
+    description = "Cython PyParsing implementation";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/cPyparsing/default.nix:18]: Missing maintainers.

File: pkgs/development/python-modules/coconut/default.nix [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  propagatedBuildInputs = [ cPyparsing pygments prompt_toolkit ipykernel mypy watchdog ];
+
+  prePatch = "find . -name '*.py' -print0 | xargs -0 dos2unix";
+  patches = [
+    (fetchpatch {
+      name = "fix-setuptools-version-check.patch";
+      url = "https://github.com/LibreCybernetics/coconut/commit/2916a087da1e063cc4438b68d4077347fd1ea4a2.patch";
+      sha256 = "136jbd2rvnifw30y73vv667002nf7sbkm5qyihshj4db7ngysr6q";
+    })
+  ];
```
Comment (SuperSandro2000) [inline pkgs/development/python-modules/coconut/default.nix:37]: Missing meta section.

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
And you are the module in the top level file ``python-packages.nix`` and the application in  ``all-packages.nix``.

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 1 above)
Comment (fabianhjr) [reply pkgs/development/python-modules/cPyparsing/default.nix:5]: Dunno, that is how it is in upstream.

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 2 above)
Comment (fabianhjr) [reply pkgs/development/python-modules/cPyparsing/default.nix:12]: Cython isn't, this builds some native extension that is used and propagated by coconut.

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 3 above)
Comment (fabianhjr) [reply pkgs/development/python-modules/cPyparsing/default.nix:18]: Should I add myself? I have not familiarized myself with maintainer-ship policy.

File: pkgs/development/python-modules/coconut/default.nix (hunk 4 above)
Comment (fabianhjr) [reply pkgs/development/python-modules/coconut/default.nix:37]: Ah right, forgot to commit, will amend shortly.

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -2827,6 +2827,12 @@
       fingerprint = "50B7 11F4 3DFD 2018 DCE6  E8D0 8A52 A140 BEBF 7D2C";
     }];
   };
+  fabianhjr = {
+    email = "fabianhjr@protonmail.com";
+    github = "fabianhjr";
+    githubId = 303897;
+    name = "Fabián Heredia Montiel";
+  };
```
Comment (jonringer) [inline maintainers/maintainer-list.nix:2835]: commit message should be:
```
maintainers: add fabianhjr
```

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 1 above)
Comment (jonringer) [inline pkgs/development/python-modules/cPyparsing/default.nix:5]: this should be a separate commit, as it's a different package addition
```
python3Packages.cpyparsing: init at 2.4.5.0.1.1
```

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 2 above)
Comment (jonringer) [inline pkgs/development/python-modules/cPyparsing/default.nix:12]:
```suggestion
  nativeBuildInputs = [ cython ];
```

File: pkgs/top-level/python-packages.nix [hunk 6]
```diff
@@ -1347,6 +1349,8 @@ in {
 
   cppy = callPackage ../development/python-modules/cppy { };
 
+  cPyparsing = callPackage ../development/python-modules/cPyparsing { };
```
Comment (jonringer) [inline pkgs/top-level/python-packages.nix:1352]:
```suggestion
  cpyparsing = callPackage ../development/python-modules/cpyparsing { };
```
Please normalize the attribute name. According to the [python contributing guidelines](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#contributing-guidelines), package names within nix should follow [PEP 0503](https://www.python.org/dev/peps/pep-0503/#normalized-names)

**Comment by fabianhjr** [comment]
Thanks for your review @jonringer, I have applied all the corrections you pointed out, rebuilt locally, and retested the binary in NixOS.

File: pkgs/development/python-modules/coconut/default.nix [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+  meta = with lib; {
+    homepage = "http://coconut-lang.org/";
+    description = "Simple, elegant, Pythonic functional programming.";
```
Comment (jonringer) [inline pkgs/development/python-modules/coconut/default.nix:41]:
```suggestion
    description = "Simple, elegant, Pythonic functional programming";
```

File: pkgs/development/python-modules/coconut/default.nix [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  version = "1.4.3";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "02rccpjnfxafp7csh49niba239yggp18cgjd2n9agalg71vfhlsh";
+  };
+
+  nativeBuildInputs = [ dos2unix findutils ];
+  propagatedBuildInputs = [ cpyparsing pygments prompt_toolkit ipykernel mypy watchdog ];
+
```
Comment (jonringer) [inline pkgs/development/python-modules/coconut/default.nix:29]: There are no tests included. Please try to checkout from source and check if they have unit tests, and try to run them. Unit tests give a good indication that they package has a high degree of validity and correctness given the python package set.

If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

```

----------------------------------------------------------------------
Ran 0 tests in 0.000s
```

File: pkgs/development/python-modules/cpyparsing/default.nix [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+  pname = "cPyparsing";
+  version = "2.4.5.0.1.1";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1kcr8v9yx9whbdn7c7rnfg4b6bk3agpa1wblhr2y6n93r8a4bxwy";
+  };
+
+  nativeBuildInputs = [ cython ];
+
```
Comment (jonringer) [inline pkgs/development/python-modules/cpyparsing/default.nix:13]: There are no tests included. Please try to checkout from source and check if they have unit tests, and try to run them. Unit tests give a good indication that they package has a high degree of validity and correctness given the python package set.

If tests are not available, then please use `pythonImportsCheck` to import the most important modules. This isn't as good as unit tests, but will usually give a good indication of run-time errors. Please see [pythonImportsCheck documentation](https://github.com/NixOS/nixpkgs/blob/master/doc/languages-frameworks/python.section.md#using-pythonimportscheck) for more information.

```

----------------------------------------------------------------------
Ran 0 tests in 0.000s
```

**Comment by fabianhjr** [comment]
Ok, changed to pull from github instead of PyPI, though I am having an issue with the coconut test-suite where it doesn't find the 'coconut' binary it just built in some tests and one test attempts to create some things at `~/` which is stopped by a permission error.

File: pkgs/development/python-modules/coconut/default.nix (hunk 8 above)
Comment (fabianhjr) [reply pkgs/development/python-modules/coconut/default.nix:32]: Tests added but not all are passing; looking into it and leaving this review comment open for now.

**Comment by fabianhjr** [comment]
Enabled passing tests only

**Comment by jonringer** [comment]
yea, tests which do impure things like doing git commands, should be deselected.

Really, we just want to make sure that the package is usable at runtime and there's not any assumptions broken. E.g.  shell commands not working `/bin/bash`, or version incompatibilities with dependencies.

File: pkgs/development/python-modules/cpyparsing/default.nix [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+  src = fetchFromGitHub {
+    owner = "evhub";
+    repo = pname;
+    rev = "aa8ee45daec5c55328446bad7202ab8f799ab0ce"; # No tags on repo
+    sha256 = "1mxa5q41cb0k4lkibs0d4lzh1w6kmhhdrsm0w0r1m3s80m05ffmw";
+  };
+
+  nativeBuildInputs = [ cython ];
+
+  checkPhase = "python tests/cPyparsing_test.py";
```
Comment (jonringer) [inline pkgs/development/python-modules/cpyparsing/default.nix:16]: for pypy and alternative interpreters
```suggestion
  checkPhase = "${python.interpreter} tests/cPyparsing_test.py";
```

File: pkgs/development/python-modules/cpyparsing/default.nix [hunk 11]
```diff
@@ -0,0 +1,24 @@
+{lib, buildPythonPackage, fetchFromGitHub, cython}:
```
Comment (jonringer) [inline pkgs/development/python-modules/cpyparsing/default.nix:1]:
```suggestion
{ lib, buildPythonPackage, fetchFromGitHub, cython, python }:
```

**Review by jonringer — COMMENTED** [review]
otherwise LGTM

**Comment by fabianhjr** [comment]
Applied your last 2 suggestions, thanks for the help.

File: pkgs/development/python-modules/cPyparsing/default.nix (hunk 3 above)
Comment (SuperSandro2000) [reply pkgs/development/python-modules/cPyparsing/default.nix]: Sure go ahead.

**Comment by fabianhjr** [comment]
Hi, just bumping this PR.

**Review by jonringer — APPROVED** [review]
LGTM

Result of nixpkgs-review pr 102675 [1](https://github.com/Mic92/nixpkgs-review)<details>  <summary>5 packages built:</summary>  <ul>    <li>coconut (python38Packages.coconut)</li>    <li>python27Packages.cpyparsing</li>    <li>python37Packages.coconut</li>    <li>python37Packages.cpyparsing</li>    <li>python38Packages.cpyparsing</li>  </ul></details>

**Comment by fabianhjr** [comment]
Thanks for reviewing, merging, and more generally maintaining Nixpkgs. uwu
